src/backend/cpu/kernel/copy.hpp

```diff
@@ -119,51 +119,33 @@ struct CopyImpl<T, T>
+    }
 
-        } else {
-            for(dim_t l=0; l<dst_dims[3]; ++l) {
-                dim_t src_loff = l*src_strides[3];
-                dim_t dst_loff = l*dst_strides[3];
-
-                if (linear_end == 3) {
-                    dim_t dst_idx = dst_loff;
-                    dim_t src_idx = src_loff;
-                    std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[3]);
-
-                } else {
-                    for(dim_t k=0; k<dst_dims[2]; ++k) {
-                        dim_t src_koff = k*src_strides[2];
-                        dim_t dst_koff = k*dst_strides[2];
-
-                        if (linear_end == 2) {
-                            dim_t dst_idx = dst_koff + dst_loff;
-                            dim_t src_idx = src_koff + src_loff;
-                            std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[2]);
-
-                        } else {
-                            for(dim_t j=0; j<dst_dims[1]; ++j) {
-                                dim_t src_joff = j*src_strides[1];
-                                dim_t dst_joff = j*dst_strides[1];
-
-                                if (linear_end == 1) {
-                                    dim_t dst_idx = dst_joff + dst_koff + dst_loff;
-                                    dim_t src_idx = src_joff + src_koff + src_loff;
-                                    std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[1]);
-
-                                } else {
-                                    for(dim_t i=0; i<dst_dims[0]; ++i) {
-                                        dim_t dst_idx = i*dst_strides[0] + dst_joff + dst_koff + dst_loff;
-                                        dim_t src_idx = i*src_strides[0] + src_joff + src_koff + src_loff;
-                                        dst_ptr[dst_idx] = src_ptr[src_idx];
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            }
+    static void copy_go(
+            T * dst_ptr,       const af::dim4 & dst_strides, const af::dim4 & dst_dims,
+            T const * src_ptr, const af::dim4 & src_strides, const af::dim4 & src_dims,
+            int dim, int linear_end)
+    {
+        for(dim_t i=0; i<dst_dims[dim]; ++i) {
+
+            // 0th dimension is recursion bottom - copy element by element
+            if (dim == 0)
+                *dst_ptr = *src_ptr;
+
+            // if we are in a higher dimension, copy the entire stride if possible
+            else if (linear_end == dim)
+                std::memcpy(dst_ptr, src_ptr, sizeof(T) * src_strides[dim]);
+
+            // otherwise recurse to a lower dimenstion
+            else
+                copy_go(dst_ptr, dst_strides, dst_dims, src_ptr, src_strides, src_dims, dim - 1, linear_end);
```

Comment:
Sure, do you also want me to split the single for loop into two separate for loops? One for dimension 0 and one for higher dimensions? So that the condition is not performed in each loop. Such as:
```c++
130         // if we are in a higher dimension, copy the entire stride if possible                                                                                                                                                                                              
131         if (linear_end == dim + 1) {                                                                                                                                                                                                                                        
132             std::memcpy(dst_ptr, src_ptr, sizeof(T) * src_strides[dim] * src_dims[dim]);                                                                                                                                                                                    
133             return;                                                                                                                                                                                                                                                         
134         }                                                                                                                                                                                                                                                                   
135                                                                                                                                                                                                                                                                             
136         // 0th dimension is recursion bottom - copy element by element                                                                                                                                                                                                      
137         if (dim == 0) {                                                                                                                                                                                                                                                     
138             for(dim_t i=0; i<dst_dims[0]; ++i) {                                                                                                                                                                                                                            
139                 *dst_ptr = *src_ptr;                                                                                                                                                                                                                                        
140                 dst_ptr += dst_strides[0];                                                                                                                                                                                                                                  
141                 src_ptr += src_strides[0];                                                                                                                                                                                                                                  
142             }                                                                                                                                                                                                                                                               
143             return;                                                                                                                                                                                                                                                         
144         }                                                                                                                                                                                                                                                                   
145                                                                                                                                                                                                                                                                             
146         // otherwise recurse to a lower dimenstion                                                                                                                                                                                                                          
147         for(dim_t i=0; i<dst_dims[dim]; ++i) {                                                                                                                                                                                                                              
148             copy_go(dst_ptr, dst_strides, dst_dims, src_ptr, src_strides, src_dims, dim - 1, linear_end);                                                                                                                                                                   
149             dst_ptr += dst_strides[dim];                                                                                                                                                                                                                                    
150             src_ptr += src_strides[dim];                                                                                                                                                                                                                                    
151         }                                                                                                                                                                                                                                                                   
152     }   
```